stacker/util.py

```diff
@@ -531,10 +536,69 @@ def create_cache_directories(self):
 
     def get_package_sources(self):
         """Makes remote python packages available for local use."""
+        # Checkout S3 repositories specified in config
+        for config in self.sources.get('s3', []):
+            self.fetch_s3_package(config=config)
         # Checkout git repositories specified in config
         for config in self.sources.get('git', []):
             self.fetch_git_package(config=config)
 
+    def fetch_s3_package(self, config):
+        """Makes a remote S3 archive available for local use
+
+        Args:
+            config (:class:`stacker.config.GitPackageSource`): git config
+                dictionary
+
+        """
+        if config['key'].endswith('.tar.gz'):
+            archive_type = '.tar.gz'
```

Comment:
What about an object called something like an `Extractor` or something similar, with a simple method/interface.  Then you'd specify the extractor to be used here, and later call it's `.extract` method or something similar?

Reply:
It'd also mean that you could do away with the chain of if/elif statements and instead have a dict that maps suffixes to extractors:

```
extractor_map = {".tar.gz": TarGzipExtractor, ".tar", TarExtractor, ".zip": ZipExtractor}
extractor = None
for suffix, klass in extractor_map.iteritems():
    if config["key"].endswith(suffix):
        extractor = klass
        break
```